subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/Cancellable.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.execution;
+
+/**
+ * Indicates that a task is cancellable.
+ */
+public interface Cancellable {
```

Comment:
If we're going to call this `Cancellable` then I suggest we remove references to tasks in the comments as it implies it represents anything that could be cancellable (and we might want potentially want to use it for other cancellable things).

Or is this intended to _only_ represent cancellable tasks?  In which case we should probably call it `CancellableTask`.

Reply:
Agreed, this should be task-agnostic.

Reply:
After thinking about this some more, I think we should split this up into two PRs:

1. Make all exec actions cancellable. This doesn't require any new API, but we can already ensure that all the correct behavior is there (e.g. task is rerun the next time etc.)
2. Add a public API. Not sure if it should be an interface you implement or rather rather a callback-based service like `CancellationToken`, as that gives users more flexibility in how they structure their code.

Reply:
Good idea. After thinking a lot, I'm going to only implement `Cancellable` API in this PR. As the first step, this PR only contains a general task cancellation (not including `Exec` cancellation). I'm not going to give users "more flexibility" in this PR, for example, should this action be able to react to cancellation?

```
class MyAction implements Action, Cancellable {
     void execute(Task t){ ... }
     void cancel() { ... }
}

task t {
     doLast(new MyAction())
}
```

Reply:
Exec cancellations are supposed to be next PR.